### ISSUE TITLE
ResourceNotFound (404) error reported creating azurerm_storage_container on account that was just recreated

### ISSUE THREAD

**Issue by tareks** [hashicorp/terraform-provider-azurerm#13070]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v1.0.5
on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.72.0
```

### Affected Resource(s)

* `azurerm_storage_container`

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "app" {
  name                = "storageaccount1234"
  resource_group_name = "test"
  location            = "australiaeast"

  account_kind             = "StorageV2"
  account_tier             = "Standard"
  account_replication_type = "LRS"
  allow_blob_public_access = false
  min_tls_version          = "TLS1_2"
}

resource "azurerm_storage_container" "app" {
  name                  = "container"
  storage_account_name  = azurerm_storage_account.app.name
  container_access_type = "private"
}

provider "azurerm" {

  features {}
}
```

### Expected Behaviour

Container should have been created after the storage account is created and available as per the implicit dependency used.

### Actual Behaviour

```
azurerm_storage_account.app: Creating...
azurerm_storage_account.app: Still creating... [10s elapsed]
azurerm_storage_account.app: Still creating... [20s elapsed]
azurerm_storage_account.app: Creation complete after 27s [id=/subscriptions/<redacted>/resourceGroups/test/providers/Microsoft.Storage/storageAccounts/<redacted>]
azurerm_storage_container.app: Creating...
╷
│ Error: failed creating container: failed creating container: containers.Client#Create: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The specified resource does not exist.\nRequestId:<redacted>\nTime:2021-08-20T00:41:53.4215244Z"
│
│   with azurerm_storage_container.app,
│   on main.tf line 13, in resource "azurerm_storage_container" "app":
│   13: resource "azurerm_storage_container" "app" {
│
```

Have also tried a data source as follows with the same results:

```hcl
data "azurerm_storage_account" "app" {
  name                = azurerm_storage_account.app.name
  resource_group_name = azurerm_resource_group.app.name
}

resource "azurerm_storage_container" "app" {
  name                  = "container"
  storage_account_name  = data.azurerm_storage_account.app.name
  container_access_type = "private"
}
```

### Steps to Reproduce

1. `terraform apply`
2. `terraform destroy`
3. `terraform apply`

### Important Factoids

Appears to be intermittent unfortunately. Seem to fail more often than not when automated.

### References

These seem to be related:

* Issue #841
* Issue #7880

**Comment by tareks** [comment]
Update: Tried adding a 5 minute wait between the destroy and apply. That still seems to not resolve the issue.

Perhaps there's a race condition due `azurerm_storage_container` is not waiting/checking for `azurerm_storage_account` to have been successfully created and just using the account name directly before the resource is created?

**Comment by magodo** [comment]
Azure storage account is discovered using DNS resolution. According to DNS protocol, each storage DNS name is cached at DNS resolvers in the path including client local DNS cache, determined by TTL (Time to Live). From the analysis above, it is probably due to the DNS endpoint of the storage account name was deleted and changed before the TTL could expire resulting in stale cache value on the client.

**Comment by magodo** [comment]
@tareks Did you also noticed that requst like `GET https://<name>.queue.core.windows.net/?comp=properties&restype=service` also failed with 404 in your case?

**Comment by axp972** [comment]
> Azure storage account is discovered using DNS resolution. According to DNS protocol, each storage DNS name is cached at DNS resolvers in the path including client local DNS cache, determined by TTL (Time to Live). From the analysis above, it is probably due to the DNS endpoint of the storage account name was deleted and changed before the TTL could expire resulting in stale cache value on the client.

I was having same issue and when I changed name of the storage account and re-tried and then it worked. I think it's as you said-  I had it created and destroyed earlier- because of this, it's DNS entry was cached and when I changed it's name it again started working.
To test this theory I again destroyed this account, then re-tried creation and hit upon exact same error.

Thanks!

**Comment by vijayakrishnar** [comment]
even, i have same issue, while updating same storage account using terraform after multiple times

**Comment by tareks** [comment]
I think as some of the comments above have already mentioned, this appears to be DNS cache related. This seems to occur infrequently for us now and have not experienced it recently, so I'll close this issue.

Usually it's just a matter of time before retrying results in success.

**Comment by dhduvall** [comment]
I've been seeing something similar, but I'm not sure if it's the same thing. In particular, while I usually am destroying and recreating the storage account with the same name, I still see the issue when there's a large gap between the two. For example, I destroyed the resource last night and when I tried to recreate it this morning, I got the error. Whatever DNS issues might be causing the problem almost certainly got resolved in the intervening 15 hours.

FWIW, I only started seeing this recently, and I've been using this resource definition for over a year. Maybe something changed on the Azure side, I dunno. Even if I was just getting lucky a lot, it was pretty much 100% of the time, whereas now it's maybe 25%, which really doesn't make it seem like luck.

I've added a random byte at the end of the name, and hopefully that'll take care of the problem (it worked the first time, at least), but I do wonder if maybe the provider or the resource could be configured to retry instead of failing? Or is there some collaboration between the Azure folks and the provider folks that might result in a cleaner solution? There are other resources that seem to have similar issues, and if it's up to me to implement the randomness workaround in each case, it'd be nice to know exactly what that set is so if I'm surprised it's only my own fault.
